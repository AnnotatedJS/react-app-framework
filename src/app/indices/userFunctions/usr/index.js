import myFunctions from "./myFunctions";

export default {
  myFunctions
};
