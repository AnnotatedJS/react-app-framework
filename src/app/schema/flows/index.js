import Auto_Forward_to_probe2 from "./Auto_Forward_to_probe2";

import Go_To_Probe2 from "./Go_To_Probe2";

import page_probe2 from "./page_probe2";

export default {
  Auto_Forward_to_probe2,

  Go_To_Probe2,

  page_probe2
};
