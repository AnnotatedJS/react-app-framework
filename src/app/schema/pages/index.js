import about from './about';
import extra from './extra';
import home from './home';

export default {
  about,
  extra,
  home,
};
