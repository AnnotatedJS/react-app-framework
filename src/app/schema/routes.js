export default [
  {
    path: '/',
    pageName: 'home',
  },
  {
    path: '/home',
    pageName: 'home',
  },
];