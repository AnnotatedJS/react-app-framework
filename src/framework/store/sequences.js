import uniqueId from 'lodash/uniqueId';
import forOwn from 'lodash/forOwn';
import get from 'lodash/get';
import merge from 'lodash/merge';
import isArray from 'lodash/isArray';
import isObject from 'lodash/isObject';
import isEmpty from 'lodash/isEmpty';
import unionWith from 'lodash/unionWith';
import * as constants from './constants';

let userFunctions = {};

function getTargetPropertiesFromEvents(events, targetProperties) {
  if (events && events.length > 0) {
    events.forEach(event => {
      const { targets } = event;
      if (targets && targets.length > 0) {
        let key;
        let propertiesObject;
        targets.forEach(target => {
          const { type, props, events } = target;
          if (type === constants.COMPONENT_TYPE && props) {
            const { componentName, componentInstance, propertyName, forwardPath } = props;
            if (propertyName) {
              key = `${componentName}_${componentInstance}`;
              propertiesObject = targetProperties[key] || {};
              // tell the that this property should be bind to the http request query
              propertiesObject[propertyName] = merge({}, propertiesObject[propertyName], {
                forwardPath
              });
              targetProperties[key] = propertiesObject;
            }
          }
          getTargetPropertiesFromEvents(events, targetProperties);
        });
      }
    });
  }
}

function deriveTargetProperties(actionSequences, targetProperties = {}) {
  forOwn(actionSequences, (value, prop) => {
    getTargetPropertiesFromEvents(value.events, targetProperties);
  });
  return targetProperties;
}

function getEventSequence(event) {
  const eventSequence = {};
  const { name, targets } = event;
  if (targets && targets.length > 0) {
    eventSequence.name = name;
    eventSequence.targets = [];
    targets.forEach(target => {
      const { type, props, events } = target;
      if (props && !isEmpty(props)) {
        const newTarget = {
          type,
        };
        if (type === constants.USER_FUNCTION_TYPE) {
          newTarget.props = {
            ...props,
            functionKey: uniqueId('function')
          };
        } else if (type === constants.COMPONENT_TYPE) {
          newTarget.props = {
            ...props,
            componentKey: uniqueId('component')
          };
        }
        if (events && events.length > 0 && type === constants.USER_FUNCTION_TYPE) {
          const newTargetEvents = [];
          events.forEach(targetEvent => {
            newTargetEvents.push(getEventSequence(targetEvent));
          });
          newTarget.events = newTargetEvents;
        }
        eventSequence.targets.push(newTarget);
      }
    });
  }
  return eventSequence;
}

function eventTargetComparator(destTarget, sourceTarget) {
  const { type: sourceType, props: sourceProps } = sourceTarget;
  const { type: destType, props: destProps } = destTarget;
  if (sourceType === destType) {
    if (sourceProps && destProps) {
      if (sourceProps.functionName && destProps.functionName) {
        return sourceProps.functionName === destProps.functionName;
      } else if (sourceProps.componentName && destProps.componentName) {
        return sourceProps.componentName === destProps.componentName
          && sourceProps.componentInstance === destProps.componentInstance
          && sourceProps.propertyName === destProps.propertyName
          && sourceProps.forwardPath === destProps.forwardPath;
      }
    }
  }
  return false;
}

function targetEventComparator(destEvent, sourceEvent) {
  return destEvent === sourceEvent || destEvent.name === sourceEvent.name;
}

function mergeEventTargets(destTargets, sourceTargets) {
  if (destTargets !== sourceTargets) {
    let resultTargets = unionWith(destTargets, sourceTargets, eventTargetComparator);
    resultTargets.forEach(resultTarget => {
      const sameSourceTarget = sourceTargets.find(sourceTarget => eventTargetComparator(resultTarget, sourceTarget));
      if (sameSourceTarget) {
        const resultTargetEvents = unionWith(resultTarget.events, sameSourceTarget.events, targetEventComparator);
        if (resultTargetEvents && resultTargetEvents.length > 0) {
          resultTargetEvents.forEach(resultTargetEvent => {
            let sameSourceTargetEvent;
            if (sameSourceTarget.events && sameSourceTarget.events.length > 0) {
              sameSourceTargetEvent =
                sameSourceTarget.events.find(sourceTargetEvent => targetEventComparator(resultTargetEvent, sourceTargetEvent));
            } else if (resultTarget.events && resultTarget.events.length > 0) {
              sameSourceTargetEvent =
                resultTarget.events.find(sourceTargetEvent => targetEventComparator(resultTargetEvent, sourceTargetEvent));
            }
            if (sameSourceTargetEvent) {
              resultTargetEvent.targets = mergeEventTargets(resultTargetEvent.targets, sameSourceTargetEvent.targets);
            }
          });
        }
        resultTarget.events = resultTargetEvents;
      }
    });
    return resultTargets;
  }
  return destTargets;
}

function getActionSequences(handlers, actionSequences = {}) {
  if (handlers && handlers.length > 0) {
    handlers.forEach(handler => {
      const { type, props, events } = handler;
      if (events && events.length > 0) {
        events.forEach(event => {
          let key;
          let handlerObject;
          if (event && event.name && event.targets && event.targets.length > 0) {
            if (type === constants.COMPONENT_TYPE) {
              key = `${props.componentName}_${props.componentInstance}`;
              handlerObject = actionSequences[key]
                || { ...props, componentKey: uniqueId('component'), events: [] };
              const eventSequence = getEventSequence(event);
              // find the same event handler name for the container
              const existingHandlerEventIndex =
                handlerObject.events.findIndex(evn => evn.name === eventSequence.name);
              if (existingHandlerEventIndex >= 0) {
                // here we should merge targets of the same container events handler
                const existingHandlerEvent = handlerObject.events[existingHandlerEventIndex];
                if (existingHandlerEvent) {
                  handlerObject.events[existingHandlerEventIndex].targets =
                    mergeEventTargets(existingHandlerEvent.targets, eventSequence.targets);
                }
              } else {
                handlerObject.events.push(eventSequence);
              }
              actionSequences[key] = handlerObject;
            }
            getActionSequences(event.targets, actionSequences);
          }
        });
      }
    });
  }
}

function createActionSequencesRecursively (handlers, actionSequences = {}) {
  forOwn(handlers, value => {
    if (isArray(value)) {
      // only arrays should be exported as flow sequences otherwise the object is assumed as a nested sequences
      getActionSequences(value, actionSequences);
    } else if (isObject(value)) {
      // if the handlers is object - it means we have a nested handlers description
      createActionSequencesRecursively(value, actionSequences);
    }
  });
  return actionSequences;
}

export function createActionSequences (handlers, functions) {
  userFunctions = functions;
  const actionSequences = createActionSequencesRecursively(handlers);
  const targetProperties = deriveTargetProperties(actionSequences);
  return { actionSequences, targetProperties };
}

export function getUserFunctionByName (functionName) {
  return get(userFunctions, functionName);
}
