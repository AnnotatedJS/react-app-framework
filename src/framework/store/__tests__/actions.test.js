import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import schema from '../../../test/actions/app/schema';
import userFunctions from '../../../test/actions/app/indices/userFunctions';
import { createActionSequences } from '../sequences';
import createContainerActions from '../actions';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

jest.setTimeout(2000 * 200);

let containerKey = 'usr.Form.Form_form';
let containerEvents;
let store;

describe('Actions', () => {

  beforeAll(() => {
    const { actionSequences } = createActionSequences(schema.flows, userFunctions);
    const containerSequence = actionSequences[containerKey];
    containerEvents = containerSequence.events;
    store = mockStore({});

    console.info = () => {};

  });

  it('run action only', () => {

    const containerActions = createContainerActions(containerKey, containerEvents);
    store.clearActions();
    store.dispatch(containerActions['onSubmit']());
    return new Promise(resolve => {
      setTimeout(() => {
        const actions = store.getActions();
        expect(actions).toMatchSnapshot();
        resolve();
      }, 2000);
    });

  });

  // it('repeat 10 times', () => {
  //   let used = process.memoryUsage().heapUsed / 1024 / 1024;
  //   console.log(`repeat 10 times before: approximately ${used} MB`);
  //
  //   let sequence = Promise.resolve();
  //   for(let i = 0; i < 10; i++) {
  //     sequence = sequence.then(() => {
  //       const containerActions = createContainerActions(containerKey, containerEvents);
  //       store.dispatch(containerActions['onSubmit']());
  //       return new Promise(resolve => {
  //         setTimeout(() => {
  //           resolve();
  //         }, 2000);
  //       })
  //     });
  //   }
  //   return sequence.then(() => {
  //     // const actions = store.getActions();
  //     // console.info('Result actions: ', JSON.stringify(actions, null, 4));
  //     used = process.memoryUsage().heapUsed / 1024 / 1024;
  //     console.log(`repeat 10 times after: approximately ${used} MB`);
  //   });
  // });

  // it('repeat 100 times', () => {
  //   let used = process.memoryUsage().heapUsed / 1024 / 1024;
  //   console.log(`repeat 100 times before: approximately ${used} MB`);
  //
  //   let sequence = Promise.resolve();
  //   for(let i = 0; i < 100; i++) {
  //     sequence = sequence.then(() => {
  //       const containerActions = createContainerActions(containerKey, containerEvents);
  //       store.dispatch(containerActions['onSubmit']());
  //       return new Promise(resolve => {
  //         setTimeout(() => {
  //           resolve();
  //         }, 2000);
  //       })
  //     });
  //   }
  //   return sequence.then(() => {
  //     // const actions = store.getActions();
  //     // console.info('Result actions: ', JSON.stringify(actions, null, 4));
  //     used = process.memoryUsage().heapUsed / 1024 / 1024;
  //     console.log(`repeat 100 times after: approximately ${used} MB`);
  //   });
  // });

});
