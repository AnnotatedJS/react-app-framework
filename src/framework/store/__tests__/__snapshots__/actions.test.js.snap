// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Actions run action only 1`] = `
Array [
  Object {
    "payload": Object {
      "notificationText": "The form was successfully validated",
    },
    "type": "usr.Form.Form_form",
  },
  Object {
    "payload": Object {
      "notificationText": "The form was successfully validated",
    },
    "type": "usr.Panel.Panel_panel",
  },
  Object {
    "payload": Object {
      "delayedObject": Object {
        "value": "Delayed object",
      },
    },
    "type": "usr.Form.Form_form",
  },
]
`;
